[Flaking Test] [sig-api-machinery] k8s.io/kubernetes/test/integration: apiserver

### Which jobs are failing?

* [sig-release-master-blocking#integration-master](https://testgrid.k8s.io/sig-release-master-blocking#integration-master&exclude-non-failed-tests=)


### Which tests are failing?

* [[sig-api-machinery] k8s.io/kubernetes/test/integration.apiserver](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-master/2064638000595210240)


### Since when has it been failing?

* First failure: Wed, 10 Jun 2026 09:17:42 UTC
* Latest failure: Wed, 10 Jun 2026 11:17:42 UTC

### Testgrid link

* [https://testgrid.k8s.io/sig-release-master-blocking#integration-master&exclude-non-failed-tests=](https://testgrid.k8s.io/sig-release-master-blocking#integration-master&exclude-non-failed-tests=) * [https://storage.googleapis.com/k8s-triage/index.html?job=ci-kubernetes-integration-master$&test=[sig-api-machinery]%20k8s.io/kubernetes/test/integration.apiserver](https://storage.googleapis.com/k8s-triage/index.html?job=ci-kubernetes-integration-master$&test=[sig-api-machinery]%20k8s.io/kubernetes/test/integration.apiserver)

### Reason for failure (if possible)

```
{Failed  === RUN   TestCustomResourceValidatorsWithSchemaConversion
    testserver.go:646: Resolved testserver package path to: "/home/prow/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
    testserver.go:433: runtime-config=map[]
    testserver.go:434: Starting kube-apiserver on port 45633...
    testserver.go:469: Waiting for /healthz to be ok...
    crd_validation_expressions_test.go:722: expect error to contain 
ule compiler initialization error: failed to convert to declType for CEL validation rules" but get: <nil>
--- FAIL: TestCustomResourceValidatorsWithSchemaConversion (5.47s)
}
```

### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig api-machinery

--- a/test/integration/apiserver/crd_validation_expressions_test.go
+++ b/test/integration/apiserver/crd_validation_expressions_test.go
@@ -22,6 +22,8 @@ import (
 	"strings"
 	"testing"
 
+	"time"
+
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
 	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
@@ -30,6 +32,7 @@ import (
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/apiserver/pkg/storage/names"
 	"k8s.io/client-go/dynamic"
 	apiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
@@ -712,14 +715,26 @@ func TestCustomResourceValidatorsWithSchemaConversion(t *testing.T) {
 		t.Fatal(err)
 	}
 	// Make an unrelated update to the previous persisted CR instance to make sure CRD handler doesn't panic
-	oldCR, err := crClient.Get(context.TODO(), name1, metav1.GetOptions{})
+	// It may take some time for the CRD schema update to be processed by the API Server's internal validation cache.
+	var lastErr error
+	err = wait.PollUntilContextTimeout(context.TODO(), 100*time.Millisecond, 10*time.Second, true, func(ctx context.Context) (bool, error) {
+		oldCR, err := crClient.Get(ctx, name1, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		oldCR.Object["metadata"].(map[string]interface{})["labels"] = map[string]interface{}{"key": "value"}
+		_, err = crClient.Update(ctx, oldCR, metav1.UpdateOptions{})
+		lastErr = err
+
+		if err != nil && strings.Contains(err.Error(), "rule compiler initialization error: failed to convert to declType for CEL validation rules") {
+			return true, nil
+		}
+
+		return false, nil
+	})
 	if err != nil {
-		t.Fatal(err)
-	}
-	oldCR.Object["metadata"].(map[string]interface{})["labels"] = map[string]interface{}{"key": "value"}
-	_, err = crClient.Update(context.TODO(), oldCR, metav1.UpdateOptions{})
-	if err == nil || !strings.Contains(err.Error(), "rule compiler initialization error: failed to convert to declType for CEL validation rules") {
-		t.Fatalf("expect error to contain \rule compiler initialization error: failed to convert to declType for CEL validation rules\" but get: %v", err)
+		t.Fatalf("expect error to contain \"rule compiler initialization error: failed to convert to declType for CEL validation rules\" but get: %v", lastErr)
 	}
 	// Create another CR instance with an array and be rejected
 	name2 := names.SimpleNameGenerator.GenerateName("cr-2")

